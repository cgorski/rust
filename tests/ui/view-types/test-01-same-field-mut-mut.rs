#![feature(view_types)]
#![allow(unused, incomplete_features)]

// TEST 1: Theorem 3 - same_field_mut_conflicts
//
// THEOREM (formalization/Core_Proven.v):
// Theorem same_field_mut_conflicts : forall f,
//   fields_conflict (mkFieldAccess f Mut) (mkFieldAccess f Mut) = true.
//
// MEANING: Two mutable accesses to the same field always conflict
//
// PROOF TECHNIQUE: Simple - same field name, both mutable â†’ conflict
//
// TEST STRATEGY:
// Create explicit mutable borrow of a field, then call view-typed method
// that also mutably accesses that field. Should error.

struct S {
    a: i32,
    b: i32,
}

impl S {
    // View spec: mutably accesses field 'a'
    fn use_a_mut(&{mut a} mut self) {
        self.a = 1;
    }
}

fn main() {
    let mut s = S { a: 0, b: 0 };

    // Explicitly borrow field 'a' mutably
    let a_ref = &mut s.a;

    // Try to call method that also mutably accesses 'a'
    // Theorem 3 guarantees this conflicts
    s.use_a_mut(); //~ ERROR cannot borrow

    // Keep the first borrow active
    *a_ref = 10;
}
