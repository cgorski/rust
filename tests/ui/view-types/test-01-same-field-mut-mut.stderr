error[E0499]: cannot borrow `s.a` as mutable more than once at a time
  --> $DIR/test-01-same-field-mut-mut.rs:38:5
   |
LL |     let a_ref = &mut s.a;
   |                 -------- first mutable borrow occurs here
...
LL |     s.use_a_mut();
   |     ^ second mutable borrow occurs here
...
LL |     *a_ref = 10;
   |     ----------- first borrow later used here

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0499`.
